File: Packages/com.github.homuler.mediapipe/Runtime/Scripts/Gpu/GlCalculatorHelper.cs
// Copyright (c) 2021 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

using System;

namespace Mediapipe
{

  public class GlCalculatorHelper : MpResourceHandle
  {
    public delegate StatusArgs NativeGlStatusFunction();
    public delegate void GlFunction();

    public GlCalculatorHelper() : base()
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__(out var ptr).Assert();
      this.ptr = ptr;
    }

    protected override void DeleteMpPtr()
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__delete(ptr);
    }

    public void InitializeForTest(GpuResources gpuResources)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__InitializeForTest__Pgr(mpPtr, gpuResources.mpPtr).Assert();

      GC.KeepAlive(gpuResources);
      GC.KeepAlive(this);
    }

    /// <param name="nativeGlStatusFunction">
    ///   Function that is run in Gl Context.
    ///   Make sure that this function doesn't throw exceptions and won't be GCed.
    /// </param>
    public void RunInGlContext(NativeGlStatusFunction nativeGlStatusFunction)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__RunInGlContext__PF(mpPtr, nativeGlStatusFunction, out var statusPtr).Assert();
      GC.KeepAlive(this);

      AssertStatusOk(statusPtr);
    }

    public void RunInGlContext(GlFunction glFunction)
    {
      RunInGlContext(() =>
      {
        try
        {
          glFunction();
          return StatusArgs.Ok();
        }
        catch (Exception e)
        {
          return StatusArgs.Internal(e.ToString());
        }
      });
    }

    public GlTexture CreateSourceTexture(ImageFrame imageFrame)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__CreateSourceTexture__Rif(mpPtr, imageFrame.mpPtr, out var texturePtr).Assert();

      GC.KeepAlive(this);
      GC.KeepAlive(imageFrame);
      return new GlTexture(texturePtr);
    }

    public GlTexture CreateSourceTexture(GpuBuffer gpuBuffer)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__CreateSourceTexture__Rgb(mpPtr, gpuBuffer.mpPtr, out var texturePtr).Assert();

      GC.KeepAlive(this);
      GC.KeepAlive(gpuBuffer);
      return new GlTexture(texturePtr);
    }

#if UNITY_IOS
    public GlTexture CreateSourceTexture(GpuBuffer gpuBuffer, int plane) {
      UnsafeNativeMethods.mp_GlCalculatorHelper__CreateSourceTexture__Rgb_i(mpPtr, gpuBuffer.mpPtr, plane, out var texturePtr).Assert();

      GC.KeepAlive(this);
      GC.KeepAlive(gpuBuffer);
      return new GlTexture(texturePtr);
    }
#endif

    public GlTexture CreateDestinationTexture(int width, int height, GpuBufferFormat format)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__CreateDestinationTexture__i_i_ui(mpPtr, width, height, format, out var texturePtr).Assert();

      GC.KeepAlive(this);
      return new GlTexture(texturePtr);
    }

    public GlTexture CreateDestinationTexture(GpuBuffer gpuBuffer)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__CreateDestinationTexture__Rgb(mpPtr, gpuBuffer.mpPtr, out var texturePtr).Assert();

      GC.KeepAlive(this);
      GC.KeepAlive(gpuBuffer);
      return new GlTexture(texturePtr);
    }

    public uint framebuffer => SafeNativeMethods.mp_GlCalculatorHelper__framebuffer(mpPtr);

    public void BindFramebuffer(GlTexture glTexture)
    {
      UnsafeNativeMethods.mp_GlCalculatorHelper__BindFrameBuffer__Rtexture(mpPtr, glTexture.mpPtr).Assert();

      GC.KeepAlive(glTexture);
      GC.KeepAlive(this);
    }

    public GlContext GetGlContext()
    {
      var glContextPtr = SafeNativeMethods.mp_GlCalculatorHelper__GetGlContext(mpPtr);

      GC.KeepAlive(this);
      return new GlContext(glContextPtr, false);
    }

    public bool Initialized()
    {
      return SafeNativeMethods.mp_GlCalculatorHelper__Initialized(mpPtr);
    }
  }
}

File: mediapipe_api/gpu/gl_calculator_helper.cc
// Copyright (c) 2021 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

#include "mediapipe_api/gpu/gl_calculator_helper.h"

#include <utility>

MpReturnCode mp_GlCalculatorHelper__(mediapipe::GlCalculatorHelper** gl_calculator_helper_out) {
  TRY
    *gl_calculator_helper_out = new mediapipe::GlCalculatorHelper();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

void mp_GlCalculatorHelper__delete(mediapipe::GlCalculatorHelper* gl_calculator_helper) { delete gl_calculator_helper; }

MpReturnCode mp_GlCalculatorHelper__InitializeForTest__Pgr(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::GpuResources* gpu_resources) {
  TRY
    gl_calculator_helper->InitializeForTest(gpu_resources);
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

MpReturnCode mp_GlCalculatorHelper__RunInGlContext__PF(mediapipe::GlCalculatorHelper* gl_calculator_helper, NativeGlStatusFunction* gl_func,
                                                       absl::Status** status_out) {
  TRY
    auto status = gl_calculator_helper->RunInGlContext([&gl_func]() -> ::absl::Status {
      return absl::Status{std::move(*(gl_func()))};
    });
    *status_out = new absl::Status{std::move(status)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

MpReturnCode mp_GlCalculatorHelper__CreateSourceTexture__Rif(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::ImageFrame* image_frame,
                                                             mediapipe::GlTexture** gl_texture_out) {
  TRY_ALL
    *gl_texture_out = new mediapipe::GlTexture{gl_calculator_helper->CreateSourceTexture(*image_frame)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

MpReturnCode mp_GlCalculatorHelper__CreateSourceTexture__Rgb(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::GpuBuffer* gpu_buffer,
                                                             mediapipe::GlTexture** gl_texture_out) {
  TRY_ALL
    *gl_texture_out = new mediapipe::GlTexture{gl_calculator_helper->CreateSourceTexture(*gpu_buffer)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

#ifdef __APPLE__
MpReturnCode mp_GlCalculatorHelper__CreateSourceTexture__Rgb_i(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::GpuBuffer* gpu_buffer, int plane,
                                                               mediapipe::GlTexture** gl_texture_out) {
  TRY_ALL
    *gl_texture_out = new mediapipe::GlTexture{gl_calculator_helper->CreateSourceTexture(*gpu_buffer, plane)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}
#endif  // __APPLE__

MpReturnCode mp_GlCalculatorHelper__CreateDestinationTexture__i_i_ui(mediapipe::GlCalculatorHelper* gl_calculator_helper, int output_width, int output_height,
                                                                     mediapipe::GpuBufferFormat format, mediapipe::GlTexture** gl_texture_out) {
  TRY
    *gl_texture_out = new mediapipe::GlTexture{gl_calculator_helper->CreateDestinationTexture(output_width, output_height, format)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

MpReturnCode mp_GlCalculatorHelper__CreateDestinationTexture__Rgb(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::GpuBuffer* gpu_buffer,
                                                                  mediapipe::GlTexture** gl_texture_out) {
  TRY_ALL
    *gl_texture_out = new mediapipe::GlTexture{gl_calculator_helper->CreateDestinationTexture(*gpu_buffer)};
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

GLuint mp_GlCalculatorHelper__framebuffer(mediapipe::GlCalculatorHelper* gl_calculator_helper) { return gl_calculator_helper->framebuffer(); }

MpReturnCode mp_GlCalculatorHelper__BindFrameBuffer__Rtexture(mediapipe::GlCalculatorHelper* gl_calculator_helper, mediapipe::GlTexture* gl_texture) {
  TRY
    gl_calculator_helper->BindFramebuffer(*gl_texture);
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

mediapipe::GlContext& mp_GlCalculatorHelper__GetGlContext(mediapipe::GlCalculatorHelper* gl_calculator_helper) { return gl_calculator_helper->GetGlContext(); }

bool mp_GlCalculatorHelper__Initialized(mediapipe::GlCalculatorHelper* gl_calculator_helper) { return gl_calculator_helper->Initialized(); }

MpReturnCode mp_GlTexture__(mediapipe::GlTexture** gl_texture_out) {
  TRY
    *gl_texture_out = new mediapipe::GlTexture();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

void mp_GlTexture__delete(mediapipe::GlTexture* gl_texture) { delete gl_texture; }

int mp_GlTexture__width(mediapipe::GlTexture* gl_texture) { return gl_texture->width(); }

int mp_GlTexture__height(mediapipe::GlTexture* gl_texture) { return gl_texture->height(); }

GLenum mp_GlTexture__target(mediapipe::GlTexture* gl_texture) { return gl_texture->target(); }

GLuint mp_GlTexture__name(mediapipe::GlTexture* gl_texture) { return gl_texture->name(); }

MpReturnCode mp_GlTexture__Release(mediapipe::GlTexture* gl_texture) {
  TRY
    gl_texture->Release();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

MpReturnCode mp_GlTexture__GetGpuBufferFrame(mediapipe::GlTexture* gl_texture, mediapipe::GpuBuffer** gpu_buffer_out) {
  TRY
    *gpu_buffer_out = gl_texture->GetFrame<mediapipe::GpuBuffer>().release();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}
